diff -up ./src/ccid.h.hp-keyboard-fix ./src/ccid.h
--- ./src/ccid.h.hp-keyboard-fix	2016-09-08 15:03:04.125684935 -0700
+++ ./src/ccid.h	2016-09-08 15:09:47.138618891 -0700
@@ -167,6 +167,8 @@ typedef struct
 #define REINER_SCT	0x0C4B0300 
 #define SEG			0x08E68000
 #define BLUDRIVEII_CCID	0x1B0E1078
+#define HPSMARTCARDTERMINAL2	0x03F0104A
+#define HPSMARTCARDTERMINAL3	0x03F0164A
 
 /*
  * The O2Micro OZ776S reader has a wrong USB descriptor
diff -up ./src/commands.c.hp-keyboard-fix ./src/commands.c
--- ./src/commands.c.hp-keyboard-fix	2016-09-08 15:02:05.101426894 -0700
+++ ./src/commands.c	2016-09-08 15:12:28.899585487 -0700
@@ -951,7 +951,9 @@ again_status:
 		/* the O2MICRO OZ 776 reader sends card absent or mute errors
 		 * when no card is inserted */
 		if (! (((OZ776 == ccid_descriptor->readerID)
-			|| (OZ776_7772 == ccid_descriptor->readerID))
+			|| (OZ776_7772 == ccid_descriptor->readerID)
+			|| (HPSMARTCARDTERMINAL2 == ccid_descriptor->readerID)
+			|| (HPSMARTCARDTERMINAL3 == ccid_descriptor->readerID))
 			&& (buffer[ERROR_OFFSET] == 0xFE)))
 #endif
 		ccid_error(buffer[ERROR_OFFSET], __FILE__, __LINE__, __FUNCTION__);    /* bError */
