diff -up ./src/ccid.c.omnikey_3022 ./src/ccid.c
--- ./src/ccid.c.omnikey_3022	2016-01-29 13:48:51.981581110 -0800
+++ ./src/ccid.c	2016-01-29 14:03:56.177214986 -0800
@@ -49,9 +49,8 @@ int ccid_open_hack(unsigned int reader_i
 	switch (ccid_descriptor->readerID)
 	{
 		case CARDMAN3121:
-			ccid_descriptor->dwNonStandardFlags = CCID_NON_STAND_OMK_3121_T1;
-			/* fall through */
 		case CARDMAN3121+1:
+			ccid_descriptor->dwNonStandardFlags = CCID_NON_STAND_OMK_3121_T1;
 			/* Reader announces APDU but is in fact TPDU */
 			ccid_descriptor->dwFeatures &= ~CCID_CLASS_EXCHANGE_MASK;
 			ccid_descriptor->dwFeatures |= CCID_CLASS_TPDU;
